### PR TITLE
Implement gRPC support in Ballerina

### LINKED ISSUE #4349: Implement gRPC support in Ballerina
**Description:**
In the current ballerina supports implementing REST services over http. But gRPC services are proved to be more faster than REST services, because it is based on http/2 and protobuf 3 and it is binary.

This task is to implement gRPC support in Ballerina such that it supports following,
- Ballerina gRPC service should be able to communicate with gRPC client created from any supported language.
- Ballerina gRPC service should support all RPC types.
- Ballerina Developer should be able to create gRPC service from either contract first approach or code first approach.

https://docs.google.com/document/d/17fwT4WK8OfV538WVSJT-_vSAPzkFn0mmM2ZgnaILVqw

**Suggested Labels (optional):**
Feature

### PR DESCRIPTION
**Pull request by Yasara123**
## Purpose
This feature introduces gRPC support in Ballerina such that it supports following,
- Ballerina gRPC service should support all RPC types.
- Ballerina Developer able to create gRPC service using Ballerina (code first approach) and the corresponding proto contract is generated at compile time automatically. Additionally, the developer has an option to create client connection at service compile time.(currently, it is done using annotation attribute “generateClientConnector” in server config) like below.
- Generating client connector from service definition.
- Supports defining message using supported primitive type and struct type.

## Goals
Implementation of the v0.9 Ballerina gRPC support.
Resolves #4349

## Approach
This feature implements the gRPC support spec:[Implement gRPC support in Ballerina](https://docs.google.com/document/d/17fwT4WK8OfV538WVSJT-_vSAPzkFn0mmM2ZgnaILVqw/edit)

## Release note
In the current ballerina supports implementing REST services over http. But gRPC services are proved to be faster than REST services because it is based on http/2 and protobuf 3 and it is binary. We are supporting gRPC following, from onward Ballerina v0.9
- Ballerina gRPC service should support all RPC types.
- Ballerina Developer should be able to create gRPC service from either contract first approach or code first approach.

## Documentation
From a user experience point of view, they will write gRPC services and connectors as same as HTTP and able to get faster transport than REST services because. More details about the gRPC protocol are available in v0.9 of the spec: [Implement gRPC support in Ballerina](https://docs.google.com/document/d/17fwT4WK8OfV538WVSJT-_vSAPzkFn0mmM2ZgnaILVqw/edit)

## Automation tests
 - Unit tests 
   > Yes
 - Integration tests
   > Yes

## Samples
- [Unary Non Blocking Stub](https://github.com/Yasara123/ballerina/tree/grpc-initial-impl/docs/ballerina-by-example/examples/grpc-unary-non-blocking-stub)
- [Unary Blocking Stub](https://github.com/Yasara123/ballerina/tree/grpc-initial-impl/docs/ballerina-by-example/examples/grpc-unary-blocking-stub)
- [Server Streaming](https://github.com/Yasara123/ballerina/tree/grpc-initial-impl/docs/ballerina-by-example/examples/grpc-server-streaming)
- [Client Streaming](https://github.com/Yasara123/ballerina/tree/grpc-initial-impl/docs/ballerina-by-example/examples/grpc-client-streaming)

### CODE REVIEW

File: docs/ballerina-by-example/examples/grpc-client-streaming/client/helloWorld.pb.bal [hunk 1]
```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2018 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chamil321) [inline docs/ballerina-by-example/examples/grpc-client-streaming/client/helloWorld.pb.bal]: We do not add licence header to BBEs. Shall we remove this?

File: docs/ballerina-by-example/examples/grpc-client-streaming/client/helloWorld.pb.bal [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+struct helloWorldStub {
+    grpc:Client clientEndpoint;
+    grpc:ServiceStub serviceStub;
+}
+function <helloWorldStub stub> initStub(grpc:Client clientEndpoint) {
+    grpc:ServiceStub navStub = {};
+    navStub.initStub(clientEndpoint, "non-blocking", descriptorKey, descriptorMap);
+    stub.serviceStub = navStub;
+}
+
```
Comment (chamil321) [inline docs/ballerina-by-example/examples/grpc-client-streaming/client/helloWorld.pb.bal]: Please remove these unnecessary lines

File: docs/ballerina-by-example/examples/grpc-client-streaming/client/helloWorld.pb.bal [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+
+
```
Comment (chamil321) [inline docs/ballerina-by-example/examples/grpc-client-streaming/client/helloWorld.pb.bal]: Shall we add docs explaining the flow?

File: docs/ballerina-by-example/examples/grpc-client-streaming/client/helloWorld.pb.bal (hunk 2 above)
Comment (Yasara123) [reply docs/ballerina-by-example/examples/grpc-client-streaming/client/helloWorld.pb.bal]: Fixed in fc9f60dd3cdaba97acb05287ea4d216112dbb632

File: docs/ballerina-by-example/examples/grpc-client-streaming/client/helloWorld.pb.bal (hunk 1 above)
Comment (Yasara123) [reply docs/ballerina-by-example/examples/grpc-client-streaming/client/helloWorld.pb.bal]: Fixed inace149f

File: stdlib/ballerina-grpc/src/main/java/org/ballerinalang/net/grpc/ssl/Parameter.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
+/*
+ *  Copyright (c) 2015 WSO2 Inc. (http://wso2.com) All Rights Reserved.
```
Comment (tishan89) [inline stdlib/ballerina-grpc/src/main/java/org/ballerinalang/net/grpc/ssl/Parameter.java]: Wrong licence year

File: stdlib/ballerina-grpc/src/main/java/org/ballerinalang/net/grpc/ssl/SSLConfig.java [hunk 5]
```diff
@@ -0,0 +1,251 @@
+/*
+ *  Copyright (c) 2015 WSO2 Inc. (http://wso2.com) All Rights Reserved.
```
Comment (tishan89) [inline stdlib/ballerina-grpc/src/main/java/org/ballerinalang/net/grpc/ssl/SSLConfig.java]: Same as above

File: stdlib/ballerina-grpc/src/main/java/org/ballerinalang/net/grpc/ssl/SSLHandlerFactory.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
+/*
+ *  Copyright (c) 2015 WSO2 Inc. (http://wso2.com) All Rights Reserved.
```
Comment (tishan89) [inline stdlib/ballerina-grpc/src/main/java/org/ballerinalang/net/grpc/ssl/SSLHandlerFactory.java]: Same as above

File: tests/ballerina-test-integration/src/test/java/org/ballerinalang/test/util/grpc/client/helloworld/HelloClient.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2017, WSO2 Inc. (http://wso2.com) All Rights Reserved.
```
Comment (tishan89) [inline tests/ballerina-test-integration/src/test/java/org/ballerinalang/test/util/grpc/client/helloworld/HelloClient.java]: Use the correct licence year

File: tests/ballerina-test-integration/src/test/java/org/ballerinalang/test/util/grpc/server/helloworld/GrpcServer.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
+/*
+ *  Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (tishan89) [inline tests/ballerina-test-integration/src/test/java/org/ballerinalang/test/util/grpc/server/helloworld/GrpcServer.java]: Same as above

File: tests/ballerina-test-integration/src/test/java/org/ballerinalang/test/util/grpc/server/helloworld/HelloWorldService.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
+/*
+ *  Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (tishan89) [inline tests/ballerina-test-integration/src/test/java/org/ballerinalang/test/util/grpc/server/helloworld/HelloWorldService.java]: Same as above

File: docs/ballerina-by-example/examples/grpc-client-streaming/client/helloWorld.pb.bal [hunk 10]
```diff
@@ -0,0 +1,57 @@
+package client;
```
[the remark this answers is not in the document]
Comment (Yasara123) [reply docs/ballerina-by-example/examples/grpc-client-streaming/client/helloWorld.pb.bal]: No, It's for site so don't put licence.

File: stdlib/ballerina-grpc/src/main/java/org/ballerinalang/net/grpc/ssl/SSLConfig.java (hunk 5 above)
Comment (Yasara123) [reply stdlib/ballerina-grpc/src/main/java/org/ballerinalang/net/grpc/ssl/SSLConfig.java]: File has removed

File: tests/ballerina-test-integration/src/test/java/org/ballerinalang/test/util/grpc/client/helloworld/HelloClient.java (hunk 7 above)
Comment (Yasara123) [reply tests/ballerina-test-integration/src/test/java/org/ballerinalang/test/util/grpc/client/helloworld/HelloClient.java]: File has removed in recent PR's

File: tests/ballerina-test-integration/src/test/java/org/ballerinalang/test/util/grpc/server/helloworld/GrpcServer.java (hunk 8 above)
Comment (Yasara123) [reply tests/ballerina-test-integration/src/test/java/org/ballerinalang/test/util/grpc/server/helloworld/GrpcServer.java]: File has removed in recent PR

File: tests/ballerina-test-integration/src/test/java/org/ballerinalang/test/util/grpc/server/helloworld/HelloWorldService.java (hunk 9 above)
Comment (Yasara123) [reply tests/ballerina-test-integration/src/test/java/org/ballerinalang/test/util/grpc/server/helloworld/HelloWorldService.java]: File has removed in recent PR

File: stdlib/ballerina-grpc/src/main/java/org/ballerinalang/net/grpc/ssl/SSLHandlerFactory.java (hunk 6 above)
Comment (Yasara123) [reply stdlib/ballerina-grpc/src/main/java/org/ballerinalang/net/grpc/ssl/SSLHandlerFactory.java]: Fixed in b83bfde0a841ea11c5fe83d8eeca9a2d5fd6764c

File: stdlib/ballerina-grpc/src/main/java/org/ballerinalang/net/grpc/ssl/SSLConfig.java (hunk 5 above)
Comment (Yasara123) [reply stdlib/ballerina-grpc/src/main/java/org/ballerinalang/net/grpc/ssl/SSLConfig.java]: Fixed in b83bfde0a841ea11c5fe83d8eeca9a2d5fd6764c

File: stdlib/ballerina-grpc/src/main/java/org/ballerinalang/net/grpc/proto/definition/Method.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+            methodBuilder.setOutputType(outputType);
+            return this;
+        }
+
+        private Builder(String methodName) {
+            methodBuilder = DescriptorProtos.MethodDescriptorProto.newBuilder();
+            methodBuilder.setName(methodName);
+        }
+
+        private DescriptorProtos.MethodDescriptorProto.Builder methodBuilder;
```
[the remark this answers is not in the document]
Comment (Yasara123) [reply stdlib/ballerina-grpc/src/main/java/org/ballerinalang/net/grpc/proto/definition/Method.java]: Fixed in https://github.com/ballerina-platform/ballerina-lang/pull/8438

**Pull request merged**